<!DOCTYPE html>
<!-- This site was created with Wowchemy. https://www.wowchemy.com -->
<!-- Last Published: September 26, 2023 --><html lang="en-us" >


<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  
  
  
    <meta name="generator" content="Wowchemy 5.7.0 for Hugo" />
  

  
  












  
  










  







  
  
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
  

  
  
  
    
      
      <link rel="preload" as="style" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap">
      <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap" media="print" onload="this.media='all'">
    
  

  
  

  
  

  

  <link rel="stylesheet" href="/css/vendor-bundle.min.047268c6dd09ad74ba54a0ba71837064.css" media="print" onload="this.media='all'">

  
  
  
    
    

    
    
    
    
      
      
    
    
    

    
    
    

    

    
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      
        
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
  

  
  
  
  
  
  
  <link rel="stylesheet" href="/css/wowchemy.4b7280f12f476311af2218ec632c93a2.css" />

  
  
  

  
  
  
  
  
  
  
    
    
    <link rel="stylesheet" href="/css/libs/chroma/dracula.min.css" title="hl-light" media="print" onload="this.media='all'" >
    <link rel="stylesheet" href="/css/libs/chroma/dracula.min.css" title="hl-dark" media="print" onload="this.media='all'" disabled>
  

  
  


























  
  
  






  <meta name="author" content="Xiaotao Shen" />





  

<meta name="description" content="Introduction If you want to construct databases for metid, you can use the massDatabase package. massdatabase is an R package that operates the online public databases and combines with other tools for streamlined compound annotation and pathway enrichment analysis." />



<link rel="alternate" hreflang="en-us" href="https://www.tidymass.org/docs/chapter7/3-construct_public_databases/" />
<link rel="canonical" href="https://www.tidymass.org/docs/chapter7/3-construct_public_databases/" />



  <link rel="manifest" href="/manifest.webmanifest" />



<link rel="icon" type="image/png" href="/media/icon_huf8b049b8318691de0a515d87795ef728_130772_32x32_fill_lanczos_center_3.png" />
<link rel="apple-touch-icon" type="image/png" href="/media/icon_huf8b049b8318691de0a515d87795ef728_130772_180x180_fill_lanczos_center_3.png" />

<meta name="theme-color" content="#EF525B" />










  
  






<meta property="twitter:card" content="summary" />

  <meta property="twitter:site" content="@wowchemy" />
  <meta property="twitter:creator" content="@wowchemy" />
<meta property="twitter:image" content="https://www.tidymass.org/media/icon_huf8b049b8318691de0a515d87795ef728_130772_512x512_fill_lanczos_center_3.png" />
<meta property="og:site_name" content="TidyMass" />
<meta property="og:url" content="https://www.tidymass.org/docs/chapter7/3-construct_public_databases/" />
<meta property="og:title" content="Construct public databases | TidyMass" />
<meta property="og:description" content="Introduction If you want to construct databases for metid, you can use the massDatabase package. massdatabase is an R package that operates the online public databases and combines with other tools for streamlined compound annotation and pathway enrichment analysis." /><meta property="og:image" content="https://www.tidymass.org/media/icon_huf8b049b8318691de0a515d87795ef728_130772_512x512_fill_lanczos_center_3.png" /><meta property="og:locale" content="en-us" />

  
    <meta
      property="article:published_time"
      content="2019-05-05T00:00:00&#43;01:00"
    />
  
  
    <meta property="article:modified_time" content="2019-05-05T00:00:00&#43;01:00">
  







  




  
  
  

  
  

  


  
  <title>Construct public databases | TidyMass</title>

  
  
  
  











</head>


<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" class="page-wrapper   " data-wc-page-id="0cbf0530eeb9387167a65676ae9c92ec" >

  
  
  
  
  
  
  
  
  
  <script src="/js/wowchemy-init.min.bc584d8639f6412fc553d13c23e02e8a.js"></script>

  


<aside class="search-modal" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#" aria-label="Close"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search" class="form-control"
        aria-label="Search...">
        
      </div>

      
      

      

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>



  <div class="page-header header--fixed">
  
  
  
  
  












<header>
  <nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
    <div class="container-xl">

      
      <div class="d-none d-lg-inline-flex">
        <a class="navbar-brand" href="/">TidyMass</a>
      </div>
      

      
      <button type="button" class="navbar-toggler" data-toggle="collapse"
              data-target="#navbar-content" aria-controls="navbar-content" aria-expanded="false" aria-label="Toggle navigation">
      <span><i class="fas fa-bars"></i></span>
      </button>
      

      
      <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
        <a class="navbar-brand" href="/">TidyMass</a>
      </div>
      

      
      
      <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

        
        <ul class="navbar-nav d-md-inline-flex">
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
              
              
              
                
              
              
            
          

          <li class="nav-item">
            <a class="nav-link " href="/#hero"><span>Home</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
              
              
              
                
              
              
            
          

          <li class="nav-item">
            <a class="nav-link " href="/#packages"><span>Packages</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
              
              
              
                
              
              
            
          

          <li class="nav-item">
            <a class="nav-link " href="/#team"><span>Team</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
              
              
              
                
              
              
            
          

          <li class="nav-item">
            <a class="nav-link " href="/#contact"><span>Contact</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
              
              
              
                
              
              
            
          

          <li class="nav-item">
            <a class="nav-link " href="/#contribution"><span>Contribution</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
          

          <li class="nav-item">
            <a class="nav-link  active" href="/docs/"><span>Tutorials</span></a>
          </li>

          
          

        

          
        </ul>
      </div>

      <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">

        
        
          
        

        
        
        
        <li class="nav-item">
          <a class="nav-link js-search" href="#" aria-label="Search"><i class="fas fa-search" aria-hidden="true"></i></a>
        </li>
        

        
        
        
        <li class="nav-item dropdown theme-dropdown">
          <a href="#" class="nav-link" data-toggle="dropdown" aria-haspopup="true" aria-label="Display preferences">
            <i class="fas fa-moon" aria-hidden="true"></i>
          </a>
          <div class="dropdown-menu">
            <a href="#" class="dropdown-item js-set-theme-light">
              <span>Light</span>
            </a>
            <a href="#" class="dropdown-item js-set-theme-dark">
              <span>Dark</span>
            </a>
            <a href="#" class="dropdown-item js-set-theme-auto">
              <span>Automatic</span>
            </a>
          </div>
        </li>
        

        
        

      </ul>

    </div>
  </nav>
</header>


  </div>

  <div class="page-body">
    
    
    

    




<div class="container-fluid docs">
  <div class="row flex-xl-nowrap">
    <div class="col-12 col-md-3 col-xl-2 docs-sidebar">
      <form class="docs-search d-flex align-items-center">
  <button class="btn docs-toggle d-md-none p-0 mr-md-3 w-100" type="button" data-toggle="collapse" data-target="#docs-nav" aria-controls="docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    <div class="d-flex">
      <span class="d-md-none pl-1 flex-grow-1 text-left overflow-hidden">
        
        
          Chapter 7 Metabolite annotation
        
      </span>
      <span><i class="fas fa-chevron-down"></i></span>
    </div>
  </button>

  
  <button class="form-control sidebar-search js-search d-none d-md-flex">
    <i class="fas fa-search pr-2"></i>
    <span class="sidebar-search-text">Search...</span>
    <span class="sidebar-search-shortcut">/</span>
  </button>
  
</form>

<nav class="collapse docs-links" id="docs-nav">
  
  
  
  
  
  

  
  
    

    
      

      


  
    
    
    
    
      
    
    

    
      <ul class="nav docs-sidenav">
        <li class=""><a href="/docs/"><i class="fas fa-house pr-1"></i>Preface</a></li>
    
      


  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter1/"><i class="fas fa-book pr-1"></i>Chapter 1 Installation</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter1/1-installation/">1 Install tidymass</a></li>



  <li class=""><a href="/docs/chapter1/2-update/">2 Update tidymass</a></li>



  <li class=""><a href="/docs/chapter1/3-docker_tidymass/">3 Docker version of tidymass</a></li>



  <li class=""><a href="/docs/chapter1/4-build_own_docker/">4 Build your own docker</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter2/"><i class="fas fa-book-reader pr-1"></i>Chapter 2 massDataset package</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter2/1-create_mass_dataset/">1 Create mass_dataset class</a></li>



  <li class=""><a href="/docs/chapter2/2-add_ms2/">2 Add MS2 spectra</a></li>



  <li class=""><a href="/docs/chapter2/3-mass_dataset_other_tools/">3 Create mass_dataset class from other tools</a></li>



  <li class=""><a href="/docs/chapter2/4-extract/">4 Extract data</a></li>



  <li class=""><a href="/docs/chapter2/5-processing_information/">5 Processing information</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter3/"><i class="fas fa-book-reader pr-1"></i>Chapter 3 Data format converter</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter3/1-proteowizard/">1 proteowizard</a></li>



  <li class=""><a href="/docs/chapter3/2-data_convert/">2 massconverter</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter4/"><i class="fas fa-book-reader pr-1"></i>Chapter 4 Raw data processing</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter4/1-raw_data_processing/">Raw data processing</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter5/"><i class="fas fa-book-reader pr-1"></i>Chapter 5 Explore data</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter5/1-data_explore/">Explore data</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter6/"><i class="fas fa-book-reader pr-1"></i>Chapter 6 Data cleaning</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter6/1-data_cleaning/">Data cleaning</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter7/"><i class="fas fa-book-reader pr-1"></i>Chapter 7 Metabolite annotation</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter7/1-metid_database_structure/">1 Databases for metID</a></li>



  <li class=""><a href="/docs/chapter7/2-in_house_database/">2 In-house MS2 database</a></li>



  <li class="active"><a href="/docs/chapter7/3-construct_public_databases/">3 Public databases</a></li>



  <li class=""><a href="/docs/chapter7/4-metabolite_annotation/">4 Metabolite annotation: An example</a></li>



  <li class=""><a href="/docs/chapter7/5-identify_single_peak/">5 Identify single peak</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter8/"><i class="fas fa-book-reader pr-1"></i>Chapter 8 Statistical analysis</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter8/1-statistical_analysis/">Statistical analysis</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter9/"><i class="fas fa-book-reader pr-1"></i>Chapter 9 Pathway enrichment</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter9/1-pahtway_enrichment/">Pathway enrichment</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter10/"><i class="fas fa-book-reader pr-1"></i>Chapter 10 Whole workflow using tidymass</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter10/1-whole_workflow/">Whole workflow using tidymass</a></li>

      
        </ul>
      
    

    
      </div>
    

      
    

    
      </ul>
    

  
</nav>

    </div>

    
    
    <div class="d-none d-xl-block col-xl-2 docs-toc">
      












      <ul class="nav toc-top">
        <li><a href="#" id="back_to_top" class="docs-toc-title">Contents</a></li>
      </ul>

      <nav id="TableOfContents">
  <ul>
    <li><a href="#introduction">Introduction</a></li>
    <li><a href="#install-massdatabase">Install <code>massdatabase</code></a></li>
    <li><a href="#bigg-database">BIGG database</a></li>
    <li><a href="#chebi-database">ChEBI database</a></li>
    <li><a href="#foodb-database">FooDB database</a></li>
    <li><a href="#gnps-database">GNPS database</a></li>
    <li><a href="#kegg-database">KEGG database</a></li>
    <li><a href="#lipidmaps">LipidMaps</a></li>
    <li><a href="#massbank">MassBank</a></li>
    <li><a href="#smpdb">SMPDB</a></li>
    <li><a href="#citation">Citation</a></li>
    <li><a href="#session-information">Session information</a></li>
  </ul>
</nav>

      











    </div>
    

    <main class="col-12 col-md-9 col-xl-8 py-md-3 pl-md-5 docs-content" role="main">

      <div class="docs-article-container">
        
          
  <nav class="d-none d-md-flex" aria-label="breadcrumb">
    <ol class="breadcrumb">
      
  
    
  
    
  
    
  

    <li class="breadcrumb-item">
      <a href="/">
        
          Home
        
      </a>
    </li>
  

    <li class="breadcrumb-item">
      <a href="/docs/">
        
          Preface
        
      </a>
    </li>
  

    <li class="breadcrumb-item">
      <a href="/docs/chapter7/">
        
          Chapter 7 Metabolite annotation
        
      </a>
    </li>
  

      <li class="breadcrumb-item active" aria-current="page">
        3 Public databases
      </li>
    </ol>
  </nav>




        
      </div>

      
      
      

      <div class="docs-article-container">
        
        <h1>Construct public databases</h1>

        <article class="article-style">

          
          

          <h2 id="introduction">Introduction</h2>
<p>If you want to construct databases for <code>metid</code>, you can use the <a href="https://massdatabase.tidymass.org/" target="_blank" rel="noopener">massDatabase</a> package. <code>massdatabase</code> is an R package that operates the online public databases and combines with other tools for streamlined compound annotation and pathway enrichment analysis. <code>massdatabase</code> is a flexible, simple, and powerful tool that can be installed on all platforms, allowing the users to leverage all the online public databases for biological function mining.</p>
<p>















<figure  >
  <div class="d-flex justify-content-center">
    <div class="w-100" ><img src="/docs/chapter7/figures/Figure_massdatabase.png" alt="" loading="lazy" data-zoomable /></div>
  </div></figure>
</p>
<h2 id="install-massdatabase">Install <code>massdatabase</code></h2>
<p>You can install <code>massdatabase</code> from <a href="https://gitlab.com/tidymass/massdatabase" target="_blank" rel="noopener">GitLab</a>.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">remotes</span><span class="p">)){</span>
</span></span><span class="line"><span class="cl"><span class="nf">install.packages</span><span class="p">(</span><span class="s">&#34;remotes&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span><span class="line"><span class="cl"><span class="n">remotes</span><span class="o">::</span><span class="nf">install_gitlab</span><span class="p">(</span><span class="s">&#34;jaspershen/massdatabase&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>or <a href="https://github.com/tidymass/massdatabase" target="_blank" rel="noopener">GitHub</a></p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">remotes</span><span class="o">::</span><span class="nf">install_github</span><span class="p">(</span><span class="s">&#34;tidymass/massdatabase&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>or tidymass.org</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">source</span><span class="p">(</span><span class="s">&#34;https://www.tidymass.org/tidymass-packages/install_tidymass.txt&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">install_tidymass</span><span class="p">(</span><span class="n">from</span> <span class="o">=</span> <span class="s">&#34;tidymass.org&#34;</span><span class="p">,</span> <span class="n">which_package</span> <span class="o">=</span> <span class="s">&#34;massdatabase&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h2 id="bigg-database">BIGG database</h2>
<p><a href="http://bigg.ucsd.edu/" target="_blank" rel="noopener">BIGG model</a>is a knowledge base of genome-scale metabolic network reconstructions.</p>
<p>Download the BIGG universal metabolite database:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">massdatabase</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">download_bigg_universal_metabolite</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">,</span> 
</span></span><span class="line"><span class="cl">                                   <span class="n">sleep</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</span></span></code></pre></div><p>Then read and convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">data</span> <span class="o">&lt;-</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">read_bigg_universal_metabolite</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;.&#34;</span><span class="p">)</span>
</span></span></code></pre></div><blockquote>
<p>It may take a while to download the database.</p>
</blockquote>
<p>Then convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">bigg_database</span> <span class="o">&lt;-</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">convert_bigg_universal2metid</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;.&#34;</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">bigg_database</span>
</span></span></code></pre></div><p>You can save the database to use later.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">save</span><span class="p">(</span><span class="n">bigg_database</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/bigg_database&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h2 id="chebi-database">ChEBI database</h2>
<p>Chemical Entities of Biological Interest (<a href="https://www.ebi.ac.uk/chebi/" target="_blank" rel="noopener">ChEBI</a>) is a freely available dictionary of molecular entities focused on ‘small’ chemical compounds.</p>
<p>Download the ChEBI compound database:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">massdatabase</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">download_chebi_compound</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>Then read and convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">data</span> <span class="o">&lt;-</span>
</span></span><span class="line"><span class="cl">  <span class="nf">read_chebi_compound</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading data...</span>
</span></span></code></pre></div><p>Then convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">chebi_database</span> <span class="o">&lt;-</span>
</span></span><span class="line"><span class="cl">  <span class="nf">convert_chebi2metid</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; No POS file in your /var/folders/m8/z8rq9r453dn9rk4zx4cw5_5h0000gn/T//Rtmp7oQCVk/file150b853bd6ecd</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; No NEG file in your /var/folders/m8/z8rq9r453dn9rk4zx4cw5_5h0000gn/T//Rtmp7oQCVk/file150b853bd6ecd</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading metabolite information...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading positive MS2 data...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading negative MS2 data...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Matching metabolites with MS2 spectra (positive)...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Matching metabolites with MS2 spectra (negative)...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; All done!</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">chebi_database</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -----------Base information------------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Version:2023-09-02</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Source:CHEBI</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Link:https://www.ebi.ac.uk/chebi/init.do</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Creater:Xiaotao Shen(shenxt@stanford.edu)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Without RT informtaion</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -----------Spectral information------------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 34 items of metabolite information:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Lab.ID; Compound.name; mz; RT; CAS.ID; HMDB.ID; KEGG.ID; Formula; mz.pos; mz.neg (top10)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 152379 metabolites in total.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 0 metabolites with spectra in positive mode.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 0 metabolites with spectra in negative mode.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Collision energy in positive mode (number:):</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Total number:0</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Collision energy in negative mode:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Total number:0</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span></code></pre></div><p>You can save the database to use later.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">save</span><span class="p">(</span><span class="n">chebi_database</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/chebi_database&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h2 id="foodb-database">FooDB database</h2>
<p><a href="https://foodb.ca/" target="_blank" rel="noopener">FooDB</a> is the world’s largest and most comprehensive resource on food constituents, chemistry and biology.</p>
<p>Download the FooDB database:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">massdatabase</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">download_foodb_compound</span><span class="p">(</span><span class="n">compound_id</span> <span class="o">=</span> <span class="s">&#34;all&#34;</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span></code></pre></div><blockquote>
<p>It may take a while to download the database.</p>
</blockquote>
<p>Then read and convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">data</span> <span class="o">&lt;-</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">read_foodb_compound</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>Then convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">foodb_dataabse</span> <span class="o">&lt;-</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">convert_foodb2metid</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction&#34;</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">foodb_dataabse</span>
</span></span></code></pre></div><p>You can save the database to use later.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">save</span><span class="p">(</span><span class="n">foodb_dataabse</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/foodb_dataabse&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h2 id="gnps-database">GNPS database</h2>
<p>The <a href="https://gnps-external.ucsd.edu/gnpslibrary" target="_blank" rel="noopener">GNPS</a> GNPS is a web-based mass spectrometry ecosystem that aims to be an open-access knowledge base for community-wide organization and sharing of raw, processed, or annotated fragmentation mass spectrometry data (MS/MS). GNPS aids in identification and discovery throughout the entire life cycle of data; from initial data acquisition/analysis to post publication.</p>
<p>Download the database:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">massdatabase</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">download_gnps_spectral_library</span><span class="p">(</span><span class="n">gnps_library</span> <span class="o">=</span> <span class="s">&#34;HMDB&#34;</span><span class="p">,</span> 
</span></span><span class="line"><span class="cl">                               <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>The argument <code>gnps_library</code> should be one of the <code>GNPS Library</code> one this website.</p>
<p><a href="https://gnps-external.ucsd.edu/gnpslibrary" target="_blank" rel="noopener">https://gnps-external.ucsd.edu/gnpslibrary</a></p>
<p>















<figure  >
  <div class="d-flex justify-content-center">
    <div class="w-100" ><img src="/docs/chapter7/figures/Screen-Shot.png" alt="" loading="lazy" data-zoomable /></div>
  </div></figure>
</p>
<blockquote>
<p>It may take a while to download the database.</p>
</blockquote>
<p>Then read and convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">data</span> <span class="o">&lt;-</span>
</span></span><span class="line"><span class="cl">  <span class="nf">read_msp_data_gnps</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/HMDB.msp&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="n">indexing</span> <span class="n">HMDB.msp</span> <span class="n">[</span><span class="o">========================================</span><span class="n">]</span> <span class="m">97.85</span><span class="n">GB</span><span class="o">/</span><span class="n">s</span><span class="p">,</span> <span class="n">eta</span><span class="o">:</span>  <span class="m">0</span><span class="n">s</span>
</span></span><span class="line"><span class="cl">                                                                                
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%</span>
</span></span></code></pre></div><p>Then convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">gnps_database</span> <span class="o">&lt;-</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">convert_gnps2metid</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Extracting MS1 inforamtion...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Done.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Extracting MS2 inforamtion...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% Done.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Organizing...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Calculating accurate mass...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Done.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% No POS file in your /var/folders/m8/z8rq9r453dn9rk4zx4cw5_5h0000gn/T//Rtmp7oQCVk/file150b8571c5a3 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; No NEG file in your /var/folders/m8/z8rq9r453dn9rk4zx4cw5_5h0000gn/T//Rtmp7oQCVk/file150b8571c5a3 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading metabolite information...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading positive MS2 data...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading negative MS2 data...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Matching metabolites with MS2 spectra (positive)...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Matching metabolites with MS2 spectra (negative)...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; All done!</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">gnps_database</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -----------Base information------------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Version:2023-09-02</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Source:GNPS</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Link:https://gnps.ucsd.edu/ProteoSAFe/static/gnps-splash.jsp</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Creater:Xiaotao Shen(shenxt@stanford.edu)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Without RT informtaion</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -----------Spectral information------------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 25 items of metabolite information:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Lab.ID; Compound.name; mz; RT; CAS.ID; HMDB.ID; KEGG.ID; Formula; mz.pos; mz.neg (top10)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 2235 metabolites in total.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1907 metabolites with spectra in positive mode.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 328 metabolites with spectra in negative mode.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Collision energy in positive mode (number:):</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Total number:1</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Unknown_1</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Collision energy in negative mode:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Total number:1</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Unknown_1</span>
</span></span></code></pre></div><p>You can save the database to use later.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">save</span><span class="p">(</span><span class="n">gnps_database</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/gnps_database&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h2 id="kegg-database">KEGG database</h2>
<p><a href="https://www.genome.jp/kegg/" target="_blank" rel="noopener">KEGG</a> is a database resource for understanding high-level functions and utilities of the biological system, such as the cell, the organism and the ecosystem, from molecular-level information, especially large-scale molecular datasets generated by genome sequencing and other high-throughput experimental technologies.</p>
<p>Download the database:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">massdatabase</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">download_kegg_compound</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span></code></pre></div><blockquote>
<p>It may take a while to download the database.</p>
</blockquote>
<p>Then read and convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">data</span> <span class="o">&lt;-</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">read_kegg_compound</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>Then convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">kegg_database</span> <span class="o">&lt;-</span>
</span></span><span class="line"><span class="cl">  <span class="nf">convert_kegg2metid</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction&#34;</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">kegg_database</span>
</span></span></code></pre></div><p>You can save the database to use later.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">save</span><span class="p">(</span><span class="n">kegg_database</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/kegg_database&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h2 id="lipidmaps">LipidMaps</h2>
<p><a href="https://www.lipidmaps.org/" target="_blank" rel="noopener">LIPID MAPS Lipidomics Gateway</a> was created in 2003 via an NIH “Glue Grant” to provide access to lipid nomenclature, databases, tools, protocols, standards, tutorials, meetings, publications, and other resources and serving the international lipid research community.</p>
<p>Download the database:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">massdatabase</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">download_lipidmaps_lipid</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction&#34;</span><span class="p">)</span>
</span></span></code></pre></div><blockquote>
<p>It may take a while to download the database.</p>
</blockquote>
<p>Uncompress the downloaded zip file.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">unzip</span><span class="p">(</span><span class="n">zipfile</span> <span class="o">=</span> <span class="s">&#34;database_construction/LMSD.sdf.zip&#34;</span><span class="p">,</span> 
</span></span><span class="line"><span class="cl">      <span class="n">exdir</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>Then read and convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">data</span> <span class="o">&lt;-</span>
</span></span><span class="line"><span class="cl">  <span class="nf">read_sdf_data_lipidmaps</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/structures.sdf&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading data, it may take a while...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Done.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Organizing...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Done.</span>
</span></span></code></pre></div><p>Then convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">lipidmaps_database</span> <span class="o">&lt;-</span>
</span></span><span class="line"><span class="cl">  <span class="nf">convert_lipidmaps2metid</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Organizing...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; No POS file in your /var/folders/m8/z8rq9r453dn9rk4zx4cw5_5h0000gn/T//RtmpCyYI3i/file15d95347e11b3</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; No NEG file in your /var/folders/m8/z8rq9r453dn9rk4zx4cw5_5h0000gn/T//RtmpCyYI3i/file15d95347e11b3</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading metabolite information...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading positive MS2 data...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Reading negative MS2 data...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Matching metabolites with MS2 spectra (positive)...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Matching metabolites with MS2 spectra (negative)...</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; All done!</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">lipidmaps_database</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -----------Base information------------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Version:2023-09-02</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Source:LIPIDMAPS</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Link:https://www.lipidmaps.org/</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Creater:Xiaotao Shen(shenxt@stanford.edu)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Without RT informtaion</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -----------Spectral information------------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 27 items of metabolite information:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Lab.ID; Compound.name; mz; RT; CAS.ID; HMDB.ID; KEGG.ID; Formula; mz.pos; mz.neg (top10)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 47500 metabolites in total.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 0 metabolites with spectra in positive mode.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 0 metabolites with spectra in negative mode.</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Collision energy in positive mode (number:):</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Total number:0</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Collision energy in negative mode:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Total number:0</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span></code></pre></div><p>You can save the database to use later.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">save</span><span class="p">(</span><span class="n">lipidmaps_database</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/lipidmaps_database&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h2 id="massbank">MassBank</h2>
<p><a href="https://massbank.eu/MassBank/" target="_blank" rel="noopener">MassBank</a> is a community effort and you are invited to contribute. Please refer to our contributor documentation and get in touch via github or email.</p>
<p>Download the database:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">massdatabase</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">download_massbank_compound</span><span class="p">(</span><span class="n">source</span> <span class="o">=</span> <span class="s">&#34;nist&#34;</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span></code></pre></div><blockquote>
<p>It may take a while to download the database.</p>
</blockquote>
<p>Then read and convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">data</span> <span class="o">&lt;-</span> <span class="nf">read_msp_data_massbank</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/massbank_compound/MassBank_NIST.msp&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%</span>
</span></span></code></pre></div><blockquote>
<p>It may take a while to read the data.</p>
</blockquote>
<p>Then convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">massbank_database</span> <span class="o">&lt;-</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">convert_massbank2metid_nist</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span></code></pre></div><blockquote>
<p>It may take a while to convert the data.</p>
</blockquote>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">massbank_database</span>
</span></span></code></pre></div><p>You can save the database to use later.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">save</span><span class="p">(</span><span class="n">massbank_database</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/massbank_database&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h2 id="smpdb">SMPDB</h2>
<p><a href="https://www.smpdb.ca/" target="_blank" rel="noopener">SMPDB</a> is an interactive, visual database containing more than 30000 small molecule pathways found in humans only.</p>
<p>Download the database:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">massdatabase</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">download_smpdb_pathway</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span></code></pre></div><blockquote>
<p>It may take a while to download the database.</p>
</blockquote>
<p>Then read and convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">data</span> <span class="o">&lt;-</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">read_smpdb_pathway</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">,</span> <span class="n">only_primarity_pathway</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</span></span></code></pre></div><p>Then convert it to <code>databaseClass</code> format.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">smpdb_pathway_database</span> <span class="o">&lt;-</span>
</span></span><span class="line"><span class="cl">  <span class="nf">convert_smpdb2metpath</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#34;database_construction/&#34;</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">smpdb_pathway_database</span>
</span></span></code></pre></div><p>You can save the database to use later.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">save</span><span class="p">(</span><span class="n">smpdb_pathway_database</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s">&#34;database_construction/smpdb_pathway_database&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h2 id="citation">Citation</h2>
<p>If you use <code>massdatabase</code> in your studies, please cite this paper:</p>
<p><a href="https://academic.oup.com/bioinformatics/article/38/19/4650/6659093" target="_blank" rel="noopener">massDatabase: utilities for the operation of the public compound and pathway database. Bioinformatics (2022).</a></p>
<h2 id="session-information">Session information</h2>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">sessionInfo</span><span class="p">()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; R version 4.3.0 (2023-04-21)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Running under: macOS Ventura 13.5.1</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Matrix products: default</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; locale:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; time zone: America/Los_Angeles</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; tzcode source: internal</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; attached base packages:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [8] base     </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; other attached packages:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [1] MSnbase_2.26.0      ProtGenerics_1.32.0 S4Vectors_0.38.1   </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [4] mzR_2.34.0          Rcpp_1.0.10         Biobase_2.60.0     </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [7] BiocGenerics_0.46.0 massdataset_1.0.25  metid_1.2.30       </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [10] metpath_1.0.8       ggplot2_3.4.2       dplyr_1.1.2        </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [13] magrittr_2.0.3      masstools_1.0.10    massdatabase_1.0.8 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; loaded via a namespace (and not attached):</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [1] RColorBrewer_1.1-3          rstudioapi_0.14            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [3] jsonlite_1.8.5              shape_1.4.6                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [5] farver_2.1.1                MALDIquant_1.22.1          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [7] rmarkdown_2.22              GlobalOptions_0.1.2        </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [9] zlibbioc_1.46.0             vctrs_0.6.2                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [11] RCurl_1.98-1.12             base64enc_0.1-3            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [13] blogdown_1.18.1             progress_1.2.2             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [15] htmltools_0.5.5             S4Arrays_1.0.4             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [17] curl_5.0.1                  cellranger_1.1.0           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [19] mzID_1.38.0                 sass_0.4.6                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [21] parallelly_1.36.0           bslib_0.5.0                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [23] htmlwidgets_1.6.2           plyr_1.8.8                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [25] impute_1.74.1               plotly_4.10.2              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [27] cachem_1.0.8                igraph_1.4.3               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [29] lifecycle_1.0.3             iterators_1.0.14           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [31] pkgconfig_2.0.3             Matrix_1.5-4               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [33] R6_2.5.1                    fastmap_1.1.1              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [35] GenomeInfoDbData_1.2.10     MatrixGenerics_1.12.2      </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [37] future_1.32.0               clue_0.3-64                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [39] digest_0.6.31               rsvg_2.4.0                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [41] pcaMethods_1.92.0           colorspace_2.1-0           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [43] furrr_0.3.1                 rprojroot_2.0.3            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [45] GenomicRanges_1.52.0        fansi_1.0.4                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [47] polyclip_1.10-4             httr_1.4.6                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [49] compiler_4.3.0              here_1.0.1                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [51] remotes_2.4.2               bit64_4.0.5                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [53] withr_2.5.0                 doParallel_1.0.17          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [55] BiocParallel_1.34.2         DBI_1.1.3                  </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [57] viridis_0.6.3               ggforce_0.4.1              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [59] MASS_7.3-58.4               ChemmineR_3.52.0           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [61] DelayedArray_0.26.3         rjson_0.2.21               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [63] tools_4.3.0                 zip_2.3.0                  </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [65] glue_1.6.2                  stringdist_0.9.10          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [67] grid_4.3.0                  cluster_2.1.4              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [69] generics_0.1.3              gtable_0.3.3               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [71] tzdb_0.4.0                  preprocessCore_1.62.1      </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [73] tidyr_1.3.0                 data.table_1.14.8          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [75] hms_1.1.3                   tidygraph_1.2.3            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [77] xml2_1.3.4                  utf8_1.2.3                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [79] XVector_0.40.0              ggrepel_0.9.3              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [81] foreach_1.5.2               pillar_1.9.0               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [83] stringr_1.5.0               vroom_1.6.3                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [85] limma_3.56.2                circlize_0.4.15            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [87] tweenr_2.0.2                lattice_0.21-8             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [89] bit_4.0.5                   tidyselect_1.2.0           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [91] ComplexHeatmap_2.16.0       Biostrings_2.68.1          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [93] pbapply_1.7-0               knitr_1.43                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [95] gridExtra_2.3               bookdown_0.34              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [97] IRanges_2.34.0              SummarizedExperiment_1.30.2</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [99] Rdisop_1.60.0               xfun_0.39                  </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [101] graphlayouts_1.0.0          matrixStats_1.0.0          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [103] DT_0.28                     stringi_1.7.12             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [105] lazyeval_0.2.2              yaml_2.3.7                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [107] evaluate_0.21               codetools_0.2-19           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [109] ggraph_2.1.0                MsCoreUtils_1.12.0         </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [111] tibble_3.2.1                BiocManager_1.30.21        </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [113] cli_3.6.1                   affyio_1.70.0              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [115] munsell_0.5.0               jquerylib_0.1.4            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [117] readxl_1.4.2                GenomeInfoDb_1.36.0        </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [119] globals_0.16.2              png_0.1-8                  </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [121] XML_3.99-0.14               parallel_4.3.0             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [123] readr_2.1.4                 prettyunits_1.1.1          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [125] bitops_1.0-7                listenv_0.9.0              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [127] viridisLite_0.4.2           scales_1.2.1               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [129] affy_1.78.0                 openxlsx_4.2.5.2           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [131] ncdf4_1.21                  purrr_1.0.1                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [133] crayon_1.5.2                GetoptLong_1.0.5           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [135] rlang_1.1.1                 rvest_1.0.3                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [137] vsn_3.68.0                  KEGGREST_1.40.0</span>
</span></span></code></pre></div>

        </article>

        





        



        
        
        <div class="article-widget">
          
<div class="post-nav">
  
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Previous</div>
    <a href="/docs/chapter7/2-in_house_database/" rel="next">Construct in-house MS2 database</a>
  </div>
  
  
  
  <div class="post-nav-item">
    <div class="meta-nav">Next</div>
    <a href="/docs/chapter7/4-metabolite_annotation/" rel="prev">Metabolite annotation: An example</a>
  </div>
  
</div>

        </div>
        
      </div>

      <div class="body-footer">
        <p>Last updated on May 5, 2019</p>

        






  

<p class="edit-page">
  <a href="https://github.com/tidymass/tidymass-website/edit/main/content/docs/chapter7/3-construct_public_databases.md">
    <i class="fas fa-pen pr-2"></i>Edit this page
  </a>
</p>



        


  
  
  

  

  
  <section id="comments">
    
<div id="disqus_thread"></div>
<script>
  var disqus_config = function () {
    
    
    
  };
  (function() {
    if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
      document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
      return;
    }
    var d = document, s = d.createElement('script'); s.async = true;
    s.src = 'https://' + "blog-at8qrhng8b" + '.disqus.com/embed.js';
    s.setAttribute('data-timestamp', +new Date());
    (d.head || d.body).appendChild(s);
  })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>


  </section>
  



        


      </div>

      <footer class="site-footer">
  <hr>

    <div style="display: flex; justify-content: space-between; padding: 20px;">
          
 <div style="flex: 1;">
            </div>
            
    
            <div style="flex: 1;">
  <p style="text-align:left;">
    Find me here:
    <br>
    <i class="fa-brands fa-github"></i> <a href="https://github.com/tidymass" target="_blank" rel="noopener">GitHub</a>
<br>
        <i class="fa-brands fa-weixin"></i> <a href="https://jaspershen.github.io/image/wechat_QR.jpg" target="_blank" rel="noopener">WeChat</a>

    <br>
        <i class="fa-brands fa-twitter"></i> <a href="https://twitter.com/xiaotaoshen1990" target="_blank" rel="noopener">Twitter</a>
        
   <br>
        <i class="fa-solid fa-envelope"></i> <a href="https://jaspershen.github.io/#contact" target="_blank" rel="noopener">Email</a>  
        
        <br>
<i class="fa-solid fa-location-dot"></i> Address: <a href="https://goo.gl/maps/jSKXmNMoj1s97uQGA" target="_blank" rel="noopener">3165 Porter Drive, Palo Alto, CA 94304</a>
   
    </p>
            </div>
            
    
            <div style="flex: 1;">
            </div>
            
    
            <div style="flex: 1;">
<script type="text/javascript" id="clstr_globe" src="//clustrmaps.com/globe.js?d=e43qUF1sR8y0B12BqEm-gDwkYxSmbro-n8KEiBINpdY"></script>
            </div>
            
    
            <div style="flex: 1;">
            </div>
    
        </div>
        
      <hr>

      











  
  
  
  
  













  
  <p class="powered-by copyright-license-text">
    © 2022-2023 Xiaotao Shen
  </p>
  




  
  
</footer>

    </main>
  </div>
</div>

  </div>

  <div class="page-footer">
    
    
  </div>

  


<script src="/js/vendor-bundle.min.f64289d8217e08e3afcd597d60836062.js"></script>




  

  
  

  









<script src="https://cdn.jsdelivr.net/npm/anchor-js@5.0.0/anchor.min.js" integrity="sha256-aQmOEF2ZD4NM/xt4hthzREIo/2PFkOX/g01WjxEV7Ys=" crossorigin="anonymous"></script>
<script>
  anchors.add();
</script>





  
  <script id="search-hit-fuse-template" type="text/x-template">
    <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
    </div>
  </script>
  
    <script src="https://cdn.jsdelivr.net/gh/krisk/Fuse@v3.2.1/dist/fuse.min.js" integrity="sha512-o38bmzBGX+hD3JHWUFCDA09btWaqrNmoJ3RXLlrysA7PP01Kgs4UlE4MhelE1v5dJR3+cxlR4qQlotsW7jKsnw==" crossorigin="anonymous"></script>
    <script src="https://cdn.jsdelivr.net/gh/julmot/mark.js@8.11.1/dist/jquery.mark.min.js" integrity="sha512-mhbv5DqBMgrWL+32MmsDOt/OAvqr/cHimk6B8y/bx/xS88MVkYGPiVv2ixKVrkywF2qHplNRUvFsAHUdxZ3Krg==" crossorigin="anonymous"></script>
  









  <script id="dsq-count-scr" src="https://blog-at8qrhng8b.disqus.com/count.js" async></script>




  
  
  
  
  
  
  

















<script id="page-data" type="application/json">{"use_headroom":false}</script>











  
  


<script src="/en/js/wowchemy.min.3322c0d94f0e691b0b24c63f4c41064b.js"></script>

























</body>
</html>
