<!DOCTYPE html>
<!-- This site was created with Wowchemy. https://www.wowchemy.com -->
<!-- Last Published: November 8, 2024 --><html lang="en-us" >


<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  
  
  
    <meta name="generator" content="Wowchemy 5.7.0 for Hugo" />
  

  
  












  
  










  







  
  
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
  

  
  
  
    
      
      <link rel="preload" as="style" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap">
      <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap" media="print" onload="this.media='all'">
    
  

  
  

  
  

  

  <link rel="stylesheet" href="/css/vendor-bundle.min.047268c6dd09ad74ba54a0ba71837064.css" media="print" onload="this.media='all'">

  
  
  
    
    

    
    
    
    
      
      
    
    
    

    
    
    

    

    
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      
        
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
  

  
  
  
  
  
  
  <link rel="stylesheet" href="/css/wowchemy.4b7280f12f476311af2218ec632c93a2.css" />

  
  
  

  
  
  
  
  
  
  
    
    
    <link rel="stylesheet" href="/css/libs/chroma/dracula.min.css" title="hl-light" media="print" onload="this.media='all'" >
    <link rel="stylesheet" href="/css/libs/chroma/dracula.min.css" title="hl-dark" media="print" onload="this.media='all'" disabled>
  

  
  


























  
  
  






  <meta name="author" content="Xiaotao Shen" />





  

<meta name="description" content="Data preparation Load the differential expressional metabolites.
library(tidymass) #&gt; Registered S3 method overwritten by &#39;Hmisc&#39;: #&gt; method from #&gt; vcov.default fit.models #&gt; ── Attaching packages ──────────────────────────────────────── tidymass 1.0.9 ── #&gt; ✔ massdataset 1." />



  <link rel="alternate" hreflang="zh" href="https://www.tidymass.org/zh/docs/chapter9/1-pahtway_enrichment/" />

<link rel="alternate" hreflang="en-us" href="https://www.tidymass.org/docs/chapter9/1-pahtway_enrichment/" />
<link rel="canonical" href="https://www.tidymass.org/docs/chapter9/1-pahtway_enrichment/" />



  <link rel="manifest" href="/manifest.webmanifest" />



<link rel="icon" type="image/png" href="/media/icon_huf8b049b8318691de0a515d87795ef728_130772_32x32_fill_lanczos_center_3.png" />
<link rel="apple-touch-icon" type="image/png" href="/media/icon_huf8b049b8318691de0a515d87795ef728_130772_180x180_fill_lanczos_center_3.png" />

<meta name="theme-color" content="#EF525B" />










  
  






<meta property="twitter:card" content="summary" />

  <meta property="twitter:site" content="@wowchemy" />
  <meta property="twitter:creator" content="@wowchemy" />
<meta property="twitter:image" content="https://www.tidymass.org/media/icon_huf8b049b8318691de0a515d87795ef728_130772_512x512_fill_lanczos_center_3.png" />
<meta property="og:site_name" content="TidyMass" />
<meta property="og:url" content="https://www.tidymass.org/docs/chapter9/1-pahtway_enrichment/" />
<meta property="og:title" content="Pathway enrichment | TidyMass" />
<meta property="og:description" content="Data preparation Load the differential expressional metabolites.
library(tidymass) #&gt; Registered S3 method overwritten by &#39;Hmisc&#39;: #&gt; method from #&gt; vcov.default fit.models #&gt; ── Attaching packages ──────────────────────────────────────── tidymass 1.0.9 ── #&gt; ✔ massdataset 1." /><meta property="og:image" content="https://www.tidymass.org/media/icon_huf8b049b8318691de0a515d87795ef728_130772_512x512_fill_lanczos_center_3.png" /><meta property="og:locale" content="en-us" />

  
    <meta
      property="article:published_time"
      content="2019-05-05T00:00:00&#43;01:00"
    />
  
  
    <meta property="article:modified_time" content="2019-05-05T00:00:00&#43;01:00">
  







  




  
  
  

  
  

  


  
  <title>Pathway enrichment | TidyMass</title>

  
  
  
  











</head>


<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" class="page-wrapper   " data-wc-page-id="7ba05ea57f0890372e082c19550a1a04" >

  
  
  
  
  
  
  
  
  
  <script src="/js/wowchemy-init.min.bc584d8639f6412fc553d13c23e02e8a.js"></script>

  


<aside class="search-modal" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#" aria-label="Close"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search" class="form-control"
        aria-label="Search...">
        
      </div>

      
      

      

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>



  <div class="page-header header--fixed">
  
  
  
  
  












<header>
  <nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
    <div class="container-xl">

      
      <div class="d-none d-lg-inline-flex">
        <a class="navbar-brand" href="/">TidyMass</a>
      </div>
      

      
      <button type="button" class="navbar-toggler" data-toggle="collapse"
              data-target="#navbar-content" aria-controls="navbar-content" aria-expanded="false" aria-label="Toggle navigation">
      <span><i class="fas fa-bars"></i></span>
      </button>
      

      
      <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
        <a class="navbar-brand" href="/">TidyMass</a>
      </div>
      

      
      
      <div class="navbar-collapse main-menu-item collapse justify-content-start" id="navbar-content">

        
        <ul class="navbar-nav d-md-inline-flex">
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
              
              
              
                
              
              
            
          

          <li class="nav-item">
            <a class="nav-link " href="/#hero"><span>Home</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
              
              
              
                
              
              
            
          

          <li class="nav-item">
            <a class="nav-link " href="/#packages"><span>Packages</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
              
              
              
                
              
              
            
          

          <li class="nav-item">
            <a class="nav-link " href="/#team"><span>Team</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
              
              
              
                
              
              
            
          

          <li class="nav-item">
            <a class="nav-link " href="/#contact"><span>Contact</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
              
              
              
                
              
              
            
          

          <li class="nav-item">
            <a class="nav-link " href="/#contribution"><span>Contribution</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
          

          <li class="nav-item">
            <a class="nav-link  active" href="/docs/"><span>Tutorials</span></a>
          </li>

          
          

        

          
        </ul>
      </div>

      <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">

        
        
          
        

        
        
        
        <li class="nav-item">
          <a class="nav-link js-search" href="#" aria-label="Search"><i class="fas fa-search" aria-hidden="true"></i></a>
        </li>
        

        
        
        
        <li class="nav-item dropdown theme-dropdown">
          <a href="#" class="nav-link" data-toggle="dropdown" aria-haspopup="true" aria-label="Display preferences">
            <i class="fas fa-moon" aria-hidden="true"></i>
          </a>
          <div class="dropdown-menu">
            <a href="#" class="dropdown-item js-set-theme-light">
              <span>Light</span>
            </a>
            <a href="#" class="dropdown-item js-set-theme-dark">
              <span>Dark</span>
            </a>
            <a href="#" class="dropdown-item js-set-theme-auto">
              <span>Automatic</span>
            </a>
          </div>
        </li>
        

        
        
        <li class="nav-item dropdown i18n-dropdown">
          <a href="#" class="nav-link " data-toggle="dropdown"
             aria-haspopup="true" aria-label="Languages">
            <i class="fas fa-globe mr-1" aria-hidden="true"></i></a>
          <div class="dropdown-menu">
            <div class="dropdown-item dropdown-item-active">
              <span>English</span>
            </div>
            
            <a class="dropdown-item" href="https://www.tidymass.org/zh/docs/chapter9/1-pahtway_enrichment/">
              <span>中文 (简体)</span>
            </a>
            
          </div>
        </li>
        

      </ul>

    </div>
  </nav>
</header>


  </div>

  <div class="page-body">
    
    
    

    




<div class="container-fluid docs">
  <div class="row flex-xl-nowrap">
    <div class="col-12 col-md-3 col-xl-2 docs-sidebar">
      <form class="docs-search d-flex align-items-center">
  <button class="btn docs-toggle d-md-none p-0 mr-md-3 w-100" type="button" data-toggle="collapse" data-target="#docs-nav" aria-controls="docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    <div class="d-flex">
      <span class="d-md-none pl-1 flex-grow-1 text-left overflow-hidden">
        
        
          Chapter 9 Pathway enrichment
        
      </span>
      <span><i class="fas fa-chevron-down"></i></span>
    </div>
  </button>

  
  <button class="form-control sidebar-search js-search d-none d-md-flex">
    <i class="fas fa-search pr-2"></i>
    <span class="sidebar-search-text">Search...</span>
    <span class="sidebar-search-shortcut">/</span>
  </button>
  
</form>

<nav class="collapse docs-links" id="docs-nav">
  
  
  
  
  
  

  
  
    

    
      

      


  
    
    
    
    
      
    
    

    
      <ul class="nav docs-sidenav">
        <li class=""><a href="/docs/"><i class="fas fa-house pr-1"></i>Preface</a></li>
    
      


  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter1/"><i class="fas fa-book pr-1"></i>Chapter 1 Installation</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter1/1-installation/">1 Install tidymass</a></li>



  <li class=""><a href="/docs/chapter1/2-update/">2 Update tidymass</a></li>



  <li class=""><a href="/docs/chapter1/3-docker_tidymass/">3 Docker version of tidymass</a></li>



  <li class=""><a href="/docs/chapter1/4-build_own_docker/">4 Build your own docker</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter2/"><i class="fas fa-book-reader pr-1"></i>Chapter 2 massDataset package</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter2/1-create_mass_dataset/">1 Create mass_dataset class</a></li>



  <li class=""><a href="/docs/chapter2/2-add_ms2/">2 Add MS2 spectra</a></li>



  <li class=""><a href="/docs/chapter2/3-mass_dataset_other_tools/">3 Create mass_dataset class from other tools</a></li>



  <li class=""><a href="/docs/chapter2/4-extract/">4 Extract data</a></li>



  <li class=""><a href="/docs/chapter2/5-processing_information/">5 Processing information</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter3/"><i class="fas fa-book-reader pr-1"></i>Chapter 3 Data format converter</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter3/1-proteowizard/">1 proteowizard</a></li>



  <li class=""><a href="/docs/chapter3/2-data_convert/">2 massconverter</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter4/"><i class="fas fa-book-reader pr-1"></i>Chapter 4 Raw data processing</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter4/1-raw_data_processing/">Raw data processing</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter5/"><i class="fas fa-book-reader pr-1"></i>Chapter 5 Explore data</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter5/1-data_explore/">Explore data</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter6/"><i class="fas fa-book-reader pr-1"></i>Chapter 6 Data cleaning</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter6/1-data_cleaning/">Data cleaning</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter7/"><i class="fas fa-book-reader pr-1"></i>Chapter 7 Metabolite annotation</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter7/1-metid_database_structure/">1 Databases for metID</a></li>



  <li class=""><a href="/docs/chapter7/2-in_house_database/">2 In-house MS2 database</a></li>



  <li class=""><a href="/docs/chapter7/3-construct_public_databases/">3 Public databases</a></li>



  <li class=""><a href="/docs/chapter7/4-public_databases/">4 Public databases download</a></li>



  <li class=""><a href="/docs/chapter7/5-metabolite_annotation/">5 Metabolite annotation: An example</a></li>



  <li class=""><a href="/docs/chapter7/6-identify_single_peak/">6 Identify single peak</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter8/"><i class="fas fa-book-reader pr-1"></i>Chapter 8 Statistical analysis</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter8/1-statistical_analysis/">Statistical analysis</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter9/"><i class="fas fa-book-reader pr-1"></i>Chapter 9 Pathway enrichment</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class="active"><a href="/docs/chapter9/1-pahtway_enrichment/">Pathway enrichment</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter10/"><i class="fas fa-book-reader pr-1"></i>Chapter 10 Whole workflow using tidymass</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter10/1-whole_workflow/">Whole workflow using tidymass</a></li>

      
        </ul>
      
    

    
      </div>
    



  
    
    
    
    
      
    
    

    
      <div class="docs-toc-item">
        <a class="docs-toc-link " href="/docs/chapter11/"><i class="fas fa-book-reader pr-1"></i>Chapter 11 Some function in tidymass</a>
    
      
        <ul class="nav docs-sidenav">
      


  <li class=""><a href="/docs/chapter11/1-formula/">Formula function</a></li>

      
        </ul>
      
    

    
      </div>
    

      
    

    
      </ul>
    

  
</nav>

    </div>

    
    
    <div class="d-none d-xl-block col-xl-2 docs-toc">
      












      <ul class="nav toc-top">
        <li><a href="#" id="back_to_top" class="docs-toc-title">Contents</a></li>
      </ul>

      <nav id="TableOfContents">
  <ul>
    <li><a href="#data-preparation">Data preparation</a></li>
    <li><a href="#pathway-enrichment">Pathway enrichment</a>
      <ul>
        <li><a href="#load-kegg-human-pathway-database">Load <code>KEGG</code> human pathway database</a></li>
      </ul>
    </li>
    <li><a href="#plot-to-show-pathway-enrichment-result">Plot to show pathway enrichment result</a></li>
    <li><a href="#session-information">Session information</a></li>
  </ul>
</nav>

      











    </div>
    

    <main class="col-12 col-md-9 col-xl-8 py-md-3 pl-md-5 docs-content" role="main">

      <div class="docs-article-container">
        
          
  <nav class="d-none d-md-flex" aria-label="breadcrumb">
    <ol class="breadcrumb">
      
  
    
  
    
  
    
  

    <li class="breadcrumb-item">
      <a href="/">
        
          Home
        
      </a>
    </li>
  

    <li class="breadcrumb-item">
      <a href="/docs/">
        
          Preface
        
      </a>
    </li>
  

    <li class="breadcrumb-item">
      <a href="/docs/chapter9/">
        
          Chapter 9 Pathway enrichment
        
      </a>
    </li>
  

      <li class="breadcrumb-item active" aria-current="page">
        Pathway enrichment
      </li>
    </ol>
  </nav>




        
      </div>

      
      
      

      <div class="docs-article-container">
        
        <h1>Pathway enrichment</h1>

        <article class="article-style">

          
          

          <h2 id="data-preparation">Data preparation</h2>
<p>Load the differential expressional metabolites.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">tidymass</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Registered S3 method overwritten by &#39;Hmisc&#39;:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   method       from      </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   vcov.default fit.models</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ── Attaching packages ──────────────────────────────────────── tidymass 1.0.9 ──</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✔ massdataset   1.0.34     ✔ metid         1.2.34</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✔ massprocesser 1.0.10     ✔ masstools     1.0.13</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✔ masscleaner   1.0.12     ✔ dplyr         1.1.4 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✔ massqc        1.0.7      ✔ ggplot2       3.5.1 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✔ massstat      1.0.6      ✔ magrittr      2.0.3 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✔ metpath       1.0.8</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✔ forcats   1.0.0     ✔ readr     2.1.5</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✔ lubridate 1.9.3     ✔ stringr   1.5.1</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✔ purrr     1.0.2     ✔ tibble    3.2.1</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ xcms::collect()          masks dplyr::collect()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ MSnbase::combine()       masks Biobase::combine(), BiocGenerics::combine(), dplyr::combine()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ tidyr::expand()          masks S4Vectors::expand()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ tidyr::extract()         masks magrittr::extract()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ metid::filter()          masks metpath::filter(), dplyr::filter(), massdataset::filter(), stats::filter()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ S4Vectors::first()       masks dplyr::first()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ xcms::groups()           masks dplyr::groups()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ dplyr::lag()             masks stats::lag()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ purrr::map()             masks mixOmics::map()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ BiocGenerics::Position() masks ggplot2::Position(), base::Position()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ purrr::reduce()          masks MSnbase::reduce()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ S4Vectors::rename()      masks dplyr::rename(), massdataset::rename()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ lubridate::second()      masks S4Vectors::second()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ lubridate::second&lt;-()    masks S4Vectors::second&lt;-()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ MASS::select()           masks dplyr::select(), massdataset::select()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ✖ purrr::set_names()       masks magrittr::set_names()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">load</span><span class="p">(</span><span class="s">&#34;statistical_analysis/object_final&#34;</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">object_final</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -------------------- </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; massdataset version: 1.0.34 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -------------------- </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1.expression_data:[ 208 x 259 data.frame]</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 2.sample_info:[ 259 x 11 data.frame]</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 259 samples:sample_06 sample_103 sample_11 ... sample_QC_38 sample_QC_39</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 3.variable_info:[ 208 x 12 data.frame]</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 208 variables:M86T95_POS M95T100_1_POS M100T160_POS ... M478T658_NEG M480T641_NEG</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 4.sample_info_note:[ 11 x 2 data.frame]</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 5.variable_info_note:[ 12 x 2 data.frame]</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 6.ms2_data:[ 2134 variables x 1939 MS2 spectra]</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -------------------- </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Processing information</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 24 processings in total</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Latest 3 processings show</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; slice_head ---------- </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;       Package Function.used                       Time</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1 massdataset  slice_head() 2024-09-25 20:59:42.450782</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 2 massdataset  slice_head()  2024-09-25 20:59:42.62459</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; mutate_fc ---------- </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;       Package Function.used                Time</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1 massdataset   mutate_fc() 2024-09-25 20:59:42</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; mutate_p_value ---------- </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;       Package    Function.used                Time</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1 massdataset mutate_p_value() 2024-09-25 20:59:42</span>
</span></span></code></pre></div><h2 id="pathway-enrichment">Pathway enrichment</h2>
<p>All the results will be placed in a folder named as <code>pathway_enrichment</code>.</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">dir.create</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s">&#34;pathway_enrichment&#34;</span><span class="p">,</span> <span class="n">showWarnings</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">diff_metabolites</span> <span class="o">&lt;-</span>
</span></span><span class="line"><span class="cl">  <span class="n">object_final</span> <span class="o">%&gt;%</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">activate_mass_dataset</span><span class="p">(</span><span class="n">what</span> <span class="o">=</span> <span class="s">&#34;variable_info&#34;</span><span class="p">)</span> <span class="o">%&gt;%</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">filter</span><span class="p">(</span><span class="n">p_value_adjust</span> <span class="o">&lt;</span> <span class="m">0.05</span><span class="p">)</span> <span class="o">%&gt;%</span> 
</span></span><span class="line"><span class="cl">  <span class="nf">extract_variable_info</span><span class="p">()</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">head</span><span class="p">(</span><span class="n">diff_metabolites</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;     variable_id        mz        rt    na_freq  na_freq.1  na_freq.2 na_freq_2</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1    M86T95_POS  86.09716  94.57264 0.02564103 0.55454545 0.29090909        NA</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 2 M95T100_1_POS  95.04975  99.77637 0.00000000 0.00000000 0.00000000        NA</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 3   M104T51_POS 104.10746  51.27993 0.00000000 0.07272727 0.00000000        NA</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 4   M113T81_POS 113.03501  80.73506 0.00000000 0.00000000 0.00000000        NA</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 5  M113T187_POS 113.06018 186.56470 0.02564103 0.01818182 0.02727273        NA</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 6   M114T52_POS 114.06658  52.31072 0.00000000 0.00000000 0.00000000        NA</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   na_freq.1_2 na_freq.2_2       fc      p_value p_value_adjust</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1          NA          NA 1.714629 3.034938e-20   4.912294e-19</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 2          NA          NA 1.439791 6.761432e-31   3.515945e-29</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 3          NA          NA 1.751085 2.695171e-08   1.334751e-07</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 4          NA          NA 1.500756 5.966911e-22   1.241118e-20</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 5          NA          NA 1.951417 1.848299e-17   2.023401e-16</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 6          NA          NA 1.382309 9.673602e-05   3.095553e-04</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;          Compound.name    CAS.ID   HMDB.ID KEGG.ID     Lab.ID Adduct  mz.error</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1           Piperidine  110-89-4      &lt;NA&gt;  C01746  MONA_2852 (M+H)+ 1.7468691</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 2               Phenol  108-95-2      &lt;NA&gt;  D01960 MONA_18506 (M+H)+ 1.4164283</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 3   5-Amino-1-pentanol 2508-29-4      &lt;NA&gt;    &lt;NA&gt;    NO07238 (M+H)+ 1.1691284</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 4               Uracil   66-22-8 HMDB00300  C00106    NO07292 (M+H)+ 1.2189636</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 5 1,4-Cyclohexanedione      &lt;NA&gt;      &lt;NA&gt;    &lt;NA&gt; MONA_14519 (M+H)+ 1.0516262</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 6           Creatinine   60-27-5 HMDB00562  C00791    NO07303 (M+H)+ 0.7404486</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   mz.match.score RT.error RT.match.score                  CE        SS</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1      0.9932417       NA             NA                  30 0.6143541</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 2      0.9955515       NA             NA                  10 0.6102452</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 3      0.9969671       NA             NA                   5 0.5971697</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 4      0.9967035       NA             NA                  10 0.6484578</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 5      0.9975454       NA             NA HCD (NCE 20-30-40%) 0.5401414</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 6      0.9987824       NA             NA                  20 0.6094410</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   Total.score   Database Level</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1   0.6996930 MoNA_0.0.1     2</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 2   0.6979976 MoNA_0.0.1     2</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 3   0.6906242 NIST_0.0.1     2</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 4   0.7213092 NIST_0.0.1     2</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 5   0.6566000 MoNA_0.0.1     2</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 6   0.6985921 NIST_0.0.1     2</span>
</span></span></code></pre></div><h3 id="load-kegg-human-pathway-database">Load <code>KEGG</code> human pathway database</h3>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">data</span><span class="p">(</span><span class="s">&#34;kegg_hsa_pathway&#34;</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s">&#34;metpath&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">kegg_hsa_pathway</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ---------Pathway source&amp;version---------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; KEGG  &amp;  2021-12-13</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -----------Pathway information------------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 345  pathways</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 334  pathways have genes</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 0  pathways have proteins</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 281  pathways have compounds</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Pathway class (top 10): Metabolism; Carbohydrate metabolism;Metabolism; Lipid metabolism</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">get_pathway_class</span><span class="p">(</span><span class="n">kegg_hsa_pathway</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; # A tibble: 43 × 2</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;    class                                                                       n</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;    &lt;chr&gt;                                                                   &lt;int&gt;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  1 Cellular Processes; Cell growth and death                                   8</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  2 Cellular Processes; Cell motility                                           1</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  3 Cellular Processes; Cellular community - eukaryotes                         5</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  4 Cellular Processes; Transport and catabolism                                7</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  5 Environmental Information Processing; Membrane transport                    1</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  6 Environmental Information Processing; Signal transduction                  26</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  7 Environmental Information Processing; Signaling molecules and interact…     5</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  8 Genetic Information Processing; Folding, sorting and degradation            7</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  9 Genetic Information Processing; Replication and repair                      7</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 10 Genetic Information Processing; Transcription                               3</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; # ℹ 33 more rows</span>
</span></span></code></pre></div><p>Remove the disease pathways:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="c1">#get the class of pathways</span>
</span></span><span class="line"><span class="cl"><span class="n">pathway_class</span> <span class="o">=</span> 
</span></span><span class="line"><span class="cl">  <span class="n">metpath</span><span class="o">::</span><span class="nf">pathway_class</span><span class="p">(</span><span class="n">kegg_hsa_pathway</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">head</span><span class="p">(</span><span class="n">pathway_class</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; $hsa00010</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [1] &#34;Metabolism; Carbohydrate metabolism&#34;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; $hsa00020</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [1] &#34;Metabolism; Carbohydrate metabolism&#34;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; $hsa00030</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [1] &#34;Metabolism; Carbohydrate metabolism&#34;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; $hsa00040</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [1] &#34;Metabolism; Carbohydrate metabolism&#34;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; $hsa00051</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [1] &#34;Metabolism; Carbohydrate metabolism&#34;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; $hsa00052</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [1] &#34;Metabolism; Carbohydrate metabolism&#34;</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">remain_idx</span> <span class="o">=</span>
</span></span><span class="line"><span class="cl">  <span class="n">pathway_class</span> <span class="o">%&gt;%</span>
</span></span><span class="line"><span class="cl">  <span class="nf">unlist</span><span class="p">()</span> <span class="o">%&gt;%</span>
</span></span><span class="line"><span class="cl">  <span class="n">stringr</span><span class="o">::</span><span class="nf">str_detect</span><span class="p">(</span><span class="s">&#34;Disease&#34;</span><span class="p">)</span> <span class="o">%&gt;%</span>
</span></span><span class="line"><span class="cl">  <span class="nf">`!`</span><span class="p">()</span> <span class="o">%&gt;%</span>
</span></span><span class="line"><span class="cl">  <span class="nf">which</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">remain_idx</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  90  91  92  93  94</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [91]  95  96  97  98  99 100 101 102 103 104 105 106 107 108 109 110 111 112</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [109] 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [127] 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [145] 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [163] 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [181] 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 201 202</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [199] 203 204 205 206 207 208 209 210 211 212 213 214 215 216 217 218 219 220</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [217] 221 222 223 224 225 226 227 228 229 230 236 240 241 242 243 244 245 246</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [235] 247 248 249 250 251 252 253 254</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">pathway_database</span> <span class="o">=</span>
</span></span><span class="line"><span class="cl">  <span class="n">kegg_hsa_pathway[remain_idx]</span>
</span></span><span class="line"><span class="cl"><span class="n">pathway_database</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ---------Pathway source&amp;version---------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; KEGG  &amp;  2021-12-13</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -----------Pathway information------------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 242  pathways</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 235  pathways have genes</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 0  pathways have proteins</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 191  pathways have compounds</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Pathway class (top 10): Metabolism; Carbohydrate metabolism;Metabolism; Lipid metabolism</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">kegg_id</span> <span class="o">&lt;-</span>
</span></span><span class="line"><span class="cl">  <span class="n">diff_metabolites</span><span class="o">$</span><span class="n">KEGG.ID</span> 
</span></span><span class="line"><span class="cl"><span class="n">kegg_id</span> <span class="o">&lt;-</span>
</span></span><span class="line"><span class="cl">  <span class="n">kegg_id[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">kegg_id</span><span class="p">)</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">kegg_id</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [1] &#34;C01746&#34; &#34;D01960&#34; &#34;C00106&#34; &#34;C00791&#34; &#34;C00906&#34; &#34;C00262&#34; &#34;C14790&#34; &#34;C02237&#34;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [9] &#34;C00073&#34; &#34;C00637&#34; &#34;C16366&#34; &#34;C07481&#34; &#34;C17846&#34; &#34;C14214&#34; &#34;C01595&#34; &#34;C00410&#34;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [17] &#34;C00762&#34; &#34;C00735&#34; &#34;C17337&#34; &#34;C01921&#34; &#34;C04230&#34; &#34;C04230&#34; &#34;C06539&#34; &#34;C00186&#34;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [25] &#34;C01546&#34; &#34;C01732&#34; &#34;C02226&#34; &#34;C00064&#34; &#34;C06104&#34; &#34;C07599&#34; &#34;0&#34;      &#34;C01601&#34;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [33] &#34;C00366&#34; &#34;C08322&#34; &#34;C16308&#34; &#34;C00294&#34; &#34;C03343&#34; &#34;C10911&#34; &#34;C05498&#34; &#34;C05472&#34;</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [41] &#34;C04555&#34;</span>
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">result</span> <span class="o">&lt;-</span> 
</span></span><span class="line"><span class="cl"><span class="nf">enrich_kegg</span><span class="p">(</span><span class="n">query_id</span> <span class="o">=</span> <span class="n">kegg_id</span><span class="p">,</span> 
</span></span><span class="line"><span class="cl">            <span class="n">query_type</span> <span class="o">=</span> <span class="s">&#34;compound&#34;</span><span class="p">,</span> 
</span></span><span class="line"><span class="cl">            <span class="n">id_type</span> <span class="o">=</span> <span class="s">&#34;KEGG&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">            <span class="n">pathway_database</span> <span class="o">=</span> <span class="n">pathway_database</span><span class="p">,</span> 
</span></span><span class="line"><span class="cl">            <span class="n">p_cutoff</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">,</span> 
</span></span><span class="line"><span class="cl">            <span class="n">p_adjust_method</span> <span class="o">=</span> <span class="s">&#34;BH&#34;</span><span class="p">,</span> 
</span></span><span class="line"><span class="cl">            <span class="n">threads</span> <span class="o">=</span> <span class="m">3</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 191 pathways.</span>
</span></span></code></pre></div><p>Check the result:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">result</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; ---------Pathway database&amp;version---------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; KEGG  &amp;  1.0.8</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -----------Enrichment result------------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 191  pathways are enriched</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 8  pathways p-values &lt; 0.05</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Steroid hormone biosynthesis</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Aldosterone-regulated sodium reabsorption</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Cushing syndrome</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Purine metabolism</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Caffeine metabolism ... (only top 5 shows)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; -----------Parameters------------</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   Package Function.used                Time</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; 1 metpath enrich_kegg() 2024-09-25 21:00:21</span>
</span></span></code></pre></div><h2 id="plot-to-show-pathway-enrichment-result">Plot to show pathway enrichment result</h2>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">enrich_bar_plot</span><span class="p">(</span><span class="n">object</span> <span class="o">=</span> <span class="n">result</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">                <span class="n">x_axis</span> <span class="o">=</span> <span class="s">&#34;p_value&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">                <span class="n">cutoff</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">)</span>
</span></span></code></pre></div><img src="/en/docs/chapter9/1-pahtway_enrichment_files/figure-html/unnamed-chunk-15-1.png" width="100%" />
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">enrich_scatter_plot</span><span class="p">(</span><span class="n">object</span> <span class="o">=</span> <span class="n">result</span><span class="p">,</span> <span class="n">y_axis</span> <span class="o">=</span> <span class="s">&#34;p_value&#34;</span><span class="p">,</span> <span class="n">y_axis_cutoff</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">)</span>
</span></span></code></pre></div><img src="/en/docs/chapter9/1-pahtway_enrichment_files/figure-html/unnamed-chunk-16-1.png" width="100%" />
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">enrich_network</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">  <span class="n">object</span> <span class="o">=</span> <span class="n">result</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">point_size</span> <span class="o">=</span> <span class="s">&#34;p_value&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">p_cutoff</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">only_significant_pathway</span> <span class="o">=</span> <span class="kc">TRUE</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span>
</span></span></code></pre></div><img src="/en/docs/chapter9/1-pahtway_enrichment_files/figure-html/unnamed-chunk-17-1.png" width="100%" />
<h2 id="session-information">Session information</h2>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">sessionInfo</span><span class="p">()</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; R version 4.4.1 (2024-06-14)</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Platform: aarch64-apple-darwin20</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Running under: macOS 15.0</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; Matrix products: default</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; locale:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; time zone: Asia/Singapore</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; tzcode source: internal</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; attached base packages:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [1] grid      stats4    stats     graphics  grDevices utils     datasets </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [8] methods   base     </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; other attached packages:</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [1] lubridate_1.9.3       forcats_1.0.0         stringr_1.5.1        </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [4] purrr_1.0.2           readr_2.1.5           tibble_3.2.1         </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [7] tidyverse_2.0.0       metid_1.2.34          metpath_1.0.8        </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [10] ComplexHeatmap_2.20.0 mixOmics_6.28.0       lattice_0.22-6       </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [13] MASS_7.3-61           massstat_1.0.6        tidyr_1.3.1          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [16] ggfortify_0.4.17      massqc_1.0.7          masscleaner_1.0.12   </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [19] MSnbase_2.30.1        ProtGenerics_1.36.0   S4Vectors_0.42.1     </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [22] Biobase_2.64.0        BiocGenerics_0.50.0   mzR_2.38.0           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [25] Rcpp_1.0.13           xcms_4.2.3            BiocParallel_1.38.0  </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [28] massprocesser_1.0.10  ggplot2_3.5.1         dplyr_1.1.4          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [31] magrittr_2.0.3        masstools_1.0.13      massdataset_1.0.34   </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [34] tidymass_1.0.9       </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; loaded via a namespace (and not attached):</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [1] fs_1.6.4                    matrixStats_1.3.0          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [3] bitops_1.0-8                fit.models_0.64            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [5] httr_1.4.7                  RColorBrewer_1.1-3         </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [7] doParallel_1.0.17           tools_4.4.1                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;   [9] doRNG_1.8.6                 backports_1.5.0            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [11] utf8_1.2.4                  R6_2.5.1                   </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [13] lazyeval_0.2.2              GetoptLong_1.0.5           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [15] withr_3.0.1                 prettyunits_1.2.0          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [17] gridExtra_2.3               preprocessCore_1.66.0      </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [19] cli_3.6.3                   fastDummies_1.7.4          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [21] labeling_0.4.3              sass_0.4.9                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [23] mvtnorm_1.3-1               robustbase_0.99-4          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [25] randomForest_4.7-1.1        proxy_0.4-27               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [27] pbapply_1.7-2               foreign_0.8-87             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [29] rrcov_1.7-6                 MetaboCoreUtils_1.12.0     </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [31] parallelly_1.38.0           itertools_0.1-3            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [33] limma_3.60.4                readxl_1.4.3               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [35] rstudioapi_0.16.0           impute_1.78.0              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [37] generics_0.1.3              shape_1.4.6.1              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [39] zip_2.3.1                   Matrix_1.7-0               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [41] MALDIquant_1.22.3           fansi_1.0.6                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [43] abind_1.4-5                 lifecycle_1.0.4            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [45] yaml_2.3.10                 SummarizedExperiment_1.34.0</span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [47] SparseArray_1.4.8           crayon_1.5.3               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [49] PSMatch_1.8.0               KEGGREST_1.44.1            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [51] pillar_1.9.0                knitr_1.48                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [53] GenomicRanges_1.56.1        rjson_0.2.22               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [55] corpcor_1.6.10              codetools_0.2-20           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [57] glue_1.7.0                  pcaMethods_1.96.0          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [59] data.table_1.16.0           remotes_2.5.0              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [61] MultiAssayExperiment_1.30.3 vctrs_0.6.5                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [63] png_0.1-8                   cellranger_1.1.0           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [65] gtable_0.3.5                cachem_1.1.0               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [67] xfun_0.47                   openxlsx_4.2.7             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [69] S4Arrays_1.4.1              tidygraph_1.3.1            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [71] pcaPP_2.0-5                 ncdf4_1.23                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [73] iterators_1.0.14            statmod_1.5.0              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [75] robust_0.7-5                progress_1.2.3             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [77] GenomeInfoDb_1.40.1         rprojroot_2.0.4            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [79] bslib_0.8.0                 affyio_1.74.0              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [81] rpart_4.1.23                colorspace_2.1-1           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [83] DBI_1.2.3                   Hmisc_5.1-3                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [85] nnet_7.3-19                 tidyselect_1.2.1           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [87] compiler_4.4.1              MassSpecWavelet_1.70.0     </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [89] htmlTable_2.4.3             DelayedArray_0.30.1        </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [91] plotly_4.10.4               bookdown_0.40              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [93] checkmate_2.3.2             scales_1.3.0               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [95] DEoptimR_1.1-3              affy_1.82.0                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [97] digest_0.6.37               rmarkdown_2.28             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt;  [99] XVector_0.44.0              htmltools_0.5.8.1          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [101] pkgconfig_2.0.3             base64enc_0.1-3            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [103] MatrixGenerics_1.16.0       highr_0.11                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [105] fastmap_1.2.0               rlang_1.1.4                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [107] GlobalOptions_0.1.2         htmlwidgets_1.6.4          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [109] UCSC.utils_1.0.0            farver_2.1.2               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [111] jquerylib_0.1.4             jsonlite_1.8.8             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [113] MsExperiment_1.6.0          mzID_1.42.0                </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [115] RCurl_1.98-1.16             Formula_1.2-5              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [117] GenomeInfoDbData_1.2.12     patchwork_1.2.0            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [119] munsell_0.5.1               viridis_0.6.5              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [121] MsCoreUtils_1.16.1          vsn_3.72.0                 </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [123] furrr_0.3.1                 stringi_1.8.4              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [125] ggraph_2.2.1                zlibbioc_1.50.0            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [127] plyr_1.8.9                  parallel_4.4.1             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [129] listenv_0.9.1               ggrepel_0.9.5              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [131] Biostrings_2.72.1           MsFeatures_1.12.0          </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [133] graphlayouts_1.1.1          hms_1.1.3                  </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [135] Spectra_1.14.1              circlize_0.4.16            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [137] igraph_2.0.3                QFeatures_1.14.2           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [139] rngtools_1.5.2              reshape2_1.4.4             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [141] XML_3.99-0.17               evaluate_0.24.0            </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [143] blogdown_1.19               BiocManager_1.30.25        </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [145] tzdb_0.4.0                  foreach_1.5.2              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [147] missForest_1.5              tweenr_2.0.3               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [149] polyclip_1.10-7             future_1.34.0              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [151] clue_0.3-65                 ggforce_0.4.2              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [153] AnnotationFilter_1.28.0     e1071_1.7-14               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [155] RSpectra_0.16-2             ggcorrplot_0.1.4.1         </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [157] viridisLite_0.4.2           class_7.3-22               </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [159] rARPACK_0.11-0              memoise_2.0.1              </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [161] ellipse_0.5.0               IRanges_2.38.1             </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [163] cluster_2.1.6               timechange_0.3.0           </span>
</span></span><span class="line"><span class="cl"><span class="c1">#&gt; [165] globals_0.16.3              here_1.0.1</span>
</span></span></code></pre></div>

        </article>

        





        



        
        
        <div class="article-widget">
          
<div class="post-nav">
  
  
</div>

        </div>
        
      </div>

      <div class="body-footer">
        <p>Last updated on May 5, 2019</p>

        






  

<p class="edit-page">
  <a href="https://github.com/tidymass/tidymass-website/edit/main/content/docs/chapter9/1-pahtway_enrichment.md">
    <i class="fas fa-pen pr-2"></i>Edit this page
  </a>
</p>



        


  
  
  

  

  
  <section id="comments">
    
<div id="disqus_thread"></div>
<script>
  var disqus_config = function () {
    
    
    
  };
  (function() {
    if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
      document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
      return;
    }
    var d = document, s = d.createElement('script'); s.async = true;
    s.src = 'https://' + "blog-at8qrhng8b" + '.disqus.com/embed.js';
    s.setAttribute('data-timestamp', +new Date());
    (d.head || d.body).appendChild(s);
  })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>


  </section>
  



        


      </div>

      <footer class="site-footer">
  <hr>

    <div style="display: flex; justify-content: space-between; padding: 20px;">
          
 <div style="flex: 1;">
            </div>
            
    
            <div style="flex: 1;">
  <p style="text-align:left;">
    Find me here:
    <br>
    <i class="fa-brands fa-github"></i> <a href="https://github.com/tidymass" target="_blank" rel="noopener">GitHub</a>
<br>
        <i class="fa-brands fa-weixin"></i> <a href="https://jaspershen.github.io/image/wechat_QR.jpg" target="_blank" rel="noopener">WeChat</a>

    <br>
        <i class="fa-brands fa-twitter"></i> <a href="https://twitter.com/xiaotaoshen1990" target="_blank" rel="noopener">Twitter</a>
        
   <br>
        <i class="fa-solid fa-envelope"></i> <a href="https://jaspershen.github.io/#contact" target="_blank" rel="noopener">Email</a>  
        
        <br>
<i class="fa-solid fa-location-dot"></i> Address: <a href="https://maps.app.goo.gl/hfCpoqgzX7zmuh3h6" target="_blank" rel="noopener">59 Nanyang Dr, Singapore 636921</a>
   
    </p>
            </div>
            
    
            <div style="flex: 1;">
            </div>
            
    
            <div style="flex: 1;">
<script type="text/javascript" id="clstr_globe" src="//clustrmaps.com/globe.js?d=e43qUF1sR8y0B12BqEm-gDwkYxSmbro-n8KEiBINpdY"></script>
            </div>
            
    
            <div style="flex: 1;">
            </div>
    
        </div>
        
      <hr>

      











  
  
  
  
  













  
  <p class="powered-by copyright-license-text">
    © 2022-2024 Shen Lab
  </p>
  




  
  
</footer>

    </main>
  </div>
</div>

  </div>

  <div class="page-footer">
    
    
  </div>

  


<script src="/js/vendor-bundle.min.938a3a7554cd9f6602290411f64d2617.js"></script>




  

  
  

  









<script src="https://cdn.jsdelivr.net/npm/anchor-js@5.0.0/anchor.min.js" integrity="sha256-aQmOEF2ZD4NM/xt4hthzREIo/2PFkOX/g01WjxEV7Ys=" crossorigin="anonymous"></script>
<script>
  anchors.add();
</script>





  
  <script id="search-hit-fuse-template" type="text/x-template">
    <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
    </div>
  </script>
  
    <script src="https://cdn.jsdelivr.net/gh/krisk/Fuse@v3.2.1/dist/fuse.min.js" integrity="sha512-o38bmzBGX+hD3JHWUFCDA09btWaqrNmoJ3RXLlrysA7PP01Kgs4UlE4MhelE1v5dJR3+cxlR4qQlotsW7jKsnw==" crossorigin="anonymous"></script>
    <script src="https://cdn.jsdelivr.net/gh/julmot/mark.js@8.11.1/dist/jquery.mark.min.js" integrity="sha512-mhbv5DqBMgrWL+32MmsDOt/OAvqr/cHimk6B8y/bx/xS88MVkYGPiVv2ixKVrkywF2qHplNRUvFsAHUdxZ3Krg==" crossorigin="anonymous"></script>
  









  <script id="dsq-count-scr" src="https://blog-at8qrhng8b.disqus.com/count.js" async></script>




  
  
  
  
  
  
  

















<script id="page-data" type="application/json">{"use_headroom":false}</script>











  
  


<script src="/en/js/wowchemy.min.85070d5fe00d43eaedff44310b81dc2c.js"></script>

























</body>
</html>
